File: workers/cron/src/pkp-api.ts
// pkp-api.ts — Typed PKP PLK Open Data API client
//
// Standard tier: 500 req/hr, 5,000 req/day
// Base URL: https://pdp-api.plk-sa.pl

const PKP_API_BASE = "https://pdp-api.plk-sa.pl";

// ---------------------------------------------------------------------------
// Types — matching Swagger DTOs
// ---------------------------------------------------------------------------

export interface OperationStationDto {
  stationId: number;
  plannedSequenceNumber: number | null;
  actualSequenceNumber: number;
  plannedArrival: string | null;
  plannedDeparture: string | null;
  actualArrival: string | null;
  actualDeparture: string | null;
  arrivalDelayMinutes?: number;
  departureDelayMinutes?: number;
  isConfirmed?: boolean;
  isCancelled?: boolean;
}

export interface TrainOperationDto {
  scheduleId: number;
  orderId: number;
  trainOrderId: number;
  operatingDate: string;
  trainStatus: string | null;
  stations: OperationStationDto[];
}

export interface OperationsResponse {
  generatedAt: string;
  pagination: {
    page: number;
    pageSize: number;
    totalCount: number;
    totalPages: number;
    hasNextPage: boolean;
  };
  trains: TrainOperationDto[];
  stations: Record<string, string>;
}

export interface StatisticsResponse {
  generatedAt: string;
  date: string;
  totalTrains: number;
  notStarted: number;
  inProgress: number;
  completed: number;
  cancelled: number;
  partialCancelled: number;
}

export interface ScheduleStationDto {
  stationId: number;
  orderNumber: number;
  arrivalTime?: string;
  departureTime?: string;
  departureTrainNumber?: string;
  arrivalTrainNumber?: string;
  arrivalCommercialCategory?: string;
  departureCommercialCategory?: string;
  arrivalPlatform?: string;
  departurePlatform?: string;
}

export interface RouteDto {
  scheduleId: number;
  orderId: number;
  trainOrderId: number;
  name: string | null;
  carrierCode: string | null;
  nationalNumber: string | null;
  commercialCategorySymbol: string | null;
  operatingDates: string[];
  stations: ScheduleStationDto[];
}

export interface SchedulesResponse {
  generatedAt: string;
  period: { from: string; to: string };
  routes: RouteDto[];
  dictionaries?: {
    stations?: Record<string, { name: string } | string>;
    carriers?: Record<string, string>;
    commercialCategories?: Record<string, string>;
  };
}

export interface DisruptionDto {
  disruptionId: number;
  disruptionTypeCode: string;
  startStation: string;
  endStation: string;
  message: string;
}

export interface DisruptionsResponse {
  success: boolean;
  data: {
    disruptions: DisruptionDto[];
  };
}

// ---------------------------------------------------------------------------
// Fetch helpers
// ---------------------------------------------------------------------------

async function pkpFetch<T>(
  path: string,
  apiKey: string,
  params?: Record<string, string>,
): Promise<T | null> {
  const url = new URL(path, PKP_API_BASE);
  if (params) {
    for (const [k, v] of Object.entries(params)) {
      url.searchParams.set(k, v);
    }
  }

  const res = await fetch(url.toString(), {
    headers: {
      "X-API-Key": apiKey,
      Accept: "application/json",
    },
  });

  if (!res.ok) {
    console.error(`PKP API error: ${res.status} ${res.statusText} for ${url.pathname}`);
    return null;
  }

  return res.json() as Promise<T>;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Fetch operations page by page, calling the handler for each page.
 * This avoids holding all 33K trains in memory at once.
 * pageSize=500 keeps each page manageable (~500 trains × ~10 stations = ~5K objects).
 */
export async function fetchOperationsPages(
  apiKey: string,
  onPage: (trains: TrainOperationDto[], stations: Record<string, string>, pageNum: number) => Promise<void>,
): Promise<{ totalTrains: number; stations: Record<string, string> }> {
  let allStations: Record<string, string> = {};
  let page = 1;
  let totalTrains = 0;
  const pageSize = 2000;

  while (true) {
    const res = await pkpFetch<OperationsResponse>(
      "/api/v1/operations",
      apiKey,
      {
        fullRoutes: "true",
        withPlanned: "true",
        page: String(page),
        pageSize: String(pageSize),
      },
    );

    if (!res || !res.trains?.length) {
      if (page === 1) {
        console.warn("[fetchOperationsPages] No data from API on first page");
      }
      break;
    }

    allStations = { ...allStations, ...res.stations };
    totalTrains += res.trains.length;

    await onPage(res.trains, allStations, page);

    if (!res.pagination.hasNextPage) break;
    page++;

    if (page > 100) {
      console.warn(`[fetchOperationsPages] Stopped at page ${page} (safety cap)`);
      break;
    }
  }

  console.log(`[fetchOperationsPages] Processed ${totalTrains} trains across ${page} pages`);
  return { totalTrains, stations: allStations };
}

export async function fetchStatistics(
  apiKey: string,
  date: string,
): Promise<StatisticsResponse | null> {
  return pkpFetch<StatisticsResponse>(
    "/api/v1/operations/statistics",
    apiKey,
    { date },
  );
}

/**
 * Fetch schedules page by page, calling the handler for each page.
 * Accumulates station + carrier dictionaries across pages and passes them to each callback.
 */
export async function fetchSchedulesPages(
  apiKey: string,
  date: string,
  onPage: (routes: RouteDto[], stations: Record<string, string>, pageNum: number) => Promise<void>,
): Promise<{ totalRoutes: number; stations: Record<string, string>; carriers: Record<string, string> }> {
  let allStations: Record<string, string> = {};
  let allCarriers: Record<string, string> = {};
  let page = 1;
  let totalRoutes = 0;
  const pageSize = 1000;

  while (true) {
    const res = await pkpFetch<SchedulesResponse>(
      "/api/v1/schedules",
      apiKey,
      {
        dateFrom: date,
        dateTo: date,
        dictionaries: "true",
        page: String(page),
        pageSize: String(pageSize),
      },
    );

    if (!res || !res.routes?.length) {
      if (page === 1) {
        console.warn("[fetchSchedulesPages] No data from API on first page");
      }
      break;
    }

    if (res.dictionaries?.stations) {
      for (const [id, info] of Object.entries(res.dictionaries.stations)) {
        allStations[id] = typeof info === 'string' ? info : info.name;
      }
    }
    if (res.dictionaries?.carriers) {
      allCarriers = { ...allCarriers, ...res.dictionaries.carriers };
    }

    totalRoutes += res.routes.length;
    await onPage(res.routes, allStations, page);

    if (res.routes.length < pageSize) break;
    page++;

    if (page > 100) {
      console.warn(`[fetchSchedulesPages] Stopped at page ${page} (safety cap)`);
      break;
    }
  }

  console.log(`[fetchSchedulesPages] Processed ${totalRoutes} routes across ${page} pages`);
  return { totalRoutes, stations: allStations, carriers: allCarriers };
}

export async function fetchDisruptions(
  apiKey: string,
): Promise<DisruptionDto[]> {
  const res = await pkpFetch<DisruptionsResponse>(
    "/api/v1/disruptions",
    apiKey,
  );

  if (!res || !res.success || !res.data?.disruptions) {
    console.warn("[fetchDisruptions] No data from API");
    return [];
  }

  return res.data.disruptions;
}

File: pipeline/pkp_api.py
"""PKP PLK Open Data API client — Python port of workers/cron/src/pkp-api.ts.

Preserves:
- 3-attempt retry with linear backoff for 5xx / network errors (no retry on 4xx)
- extract_train_number precedence: nationalNumber → internationalDepartureNumber →
  internationalArrivalNumber → name → compound {scheduleId}/{orderId}
"""
from __future__ import annotations

import time
from typing import Callable

import requests

PKP_API_BASE = "https://pdp-api.plk-sa.pl"
MAX_ATTEMPTS = 3


def pkp_fetch(
    path: str,
    api_key: str,
    params: dict[str, str] | None = None,
    timeout: float = 30.0,
) -> dict | None:
    """Fetch a PKP API path with retry. Returns parsed JSON or None on failure."""
    url = PKP_API_BASE + path
    headers = {"X-API-Key": api_key, "Accept": "application/json"}

    for attempt in range(1, MAX_ATTEMPTS + 1):
        try:
            r = requests.get(url, headers=headers, params=params or {}, timeout=timeout)
            if r.ok:
                return r.json()
            if r.status_code < 500 or attempt == MAX_ATTEMPTS:
                print(
                    f"[pkp_fetch] {r.status_code} {r.reason} for {path} "
                    f"(attempt {attempt}/{MAX_ATTEMPTS})"
                )
                return None
            print(f"[pkp_fetch] {r.status_code} on {path}, retrying")
        except requests.RequestException as e:
            if attempt == MAX_ATTEMPTS:
                print(f"[pkp_fetch] network error on {path}: {e}")
                return None
            print(f"[pkp_fetch] network error on {path}, retrying: {e}")
        time.sleep(0.5 * attempt)
    return None


def extract_train_number(route: dict) -> str:
    """Return the real train number for a /schedules route.

    Precedence matches workers/cron/src/index.ts:extractTrainNumber exactly:
    nationalNumber → internationalDepartureNumber → internationalArrivalNumber →
    name → compound {scheduleId}/{orderId} (last-resort placeholder).
    """
    for key in (
        "nationalNumber",
        "internationalDepartureNumber",
        "internationalArrivalNumber",
        "name",
    ):
        value = route.get(key)
        if value is not None and isinstance(value, str):
            stripped = value.strip()
            if stripped:
                return stripped
    return f"{route['scheduleId']}/{route['orderId']}"


def fetch_operations_pages(
    api_key: str,
    on_page: Callable[[list[dict], dict[str, str], int], None],
) -> dict:
    """Stream paged /api/v1/operations, calling on_page(trains, stations, page_num) per page.

    Returns {'total_trains': N, 'stations': {id: name}}.
    """
    all_stations: dict[str, str] = {}
    total_trains = 0
    page = 1
    page_size = 2000
    max_pages = 100

    while page <= max_pages:
        res = pkp_fetch(
            "/api/v1/operations",
            api_key,
            {
                "fullRoutes": "true",
                "withPlanned": "true",
                "page": str(page),
                "pageSize": str(page_size),
            },
        )
        if not res or not res.get("trains"):
            if page == 1:
                print("[fetch_operations_pages] no data on first page")
            break

        all_stations.update(res.get("stations", {}))
        trains = res["trains"]
        total_trains += len(trains)
        on_page(trains, all_stations, page)

        if not res.get("pagination", {}).get("hasNextPage"):
            break
        page += 1

    print(f"[fetch_operations_pages] processed {total_trains} trains across {page} pages")
    return {"total_trains": total_trains, "stations": all_stations}


def fetch_statistics(api_key: str, date: str) -> dict | None:
    """GET /api/v1/operations/statistics?date=YYYY-MM-DD"""
    return pkp_fetch("/api/v1/operations/statistics", api_key, {"date": date})


def fetch_schedules_pages(
    api_key: str,
    date: str,
    on_page: Callable[[list[dict], dict[str, str], int], None],
) -> dict:
    """Stream paged /api/v1/schedules for a single date. Accumulates dictionaries across pages."""
    all_stations: dict[str, str] = {}
    all_carriers: dict[str, str] = {}
    total_routes = 0
    page = 1
    page_size = 1000
    max_pages = 100

    while page <= max_pages:
        res = pkp_fetch(
            "/api/v1/schedules",
            api_key,
            {
                "dateFrom": date,
                "dateTo": date,
                "dictionaries": "true",
                "page": str(page),
                "pageSize": str(page_size),
            },
        )
        if not res or not res.get("routes"):
            if page == 1:
                print("[fetch_schedules_pages] no data on first page")
            break

        dictionaries = res.get("dictionaries") or {}
        stations_dict = dictionaries.get("stations") or {}
        for sid, info in stations_dict.items():
            all_stations[str(sid)] = info["name"] if isinstance(info, dict) else info
        all_carriers.update(dictionaries.get("carriers") or {})

        routes = res["routes"]
        total_routes += len(routes)
        on_page(routes, all_stations, page)

        if len(routes) < page_size:
            break
        page += 1

    return {"total_routes": total_routes, "stations": all_stations, "carriers": all_carriers}


def fetch_disruptions(api_key: str) -> list[dict]:
    """GET /api/v1/disruptions. Returns list or empty on failure."""
    res = pkp_fetch("/api/v1/disruptions", api_key)
    if not res or not res.get("success") or not res.get("data", {}).get("disruptions"):
        return []
    return res["data"]["disruptions"]
